Proposal for Contributor Experience SIG (#190)
spinnaker/governance: cx-sig-charter-draft.md

# Contributor Experience Special Interest Group Charter

This charter uses the Roles and Organization Management outlined in [sig-governance](https://github.com/spinnaker/governance/blob/master/governance.md).

## Scope

The Contributor Experience Special Interest Group (SIG) is responsible for improving the experience of those who contribute upstream to the Spinnaker project. We do this by creating, and maintaining programs and processes that promote community health and reduce project friction while retiring those programs and processes that don't. Being conscientious of our contributor base is critical to scaling the project, growing the ecosystem, and helping the project succeed.

We do this by listening - whether it’s through our roadshows to SIG meetings, surveys, data, or [GitHub issues](https://github.com/spinnaker/spinnaker/issues), we take in the feedback and turn it into our project list. We build a welcoming and inclusive community of contributors by giving them places to be heard and productive.

### In scope

#### Projects & Services

> -  In accordance with the code of conduct, establish policies, standards, and procedures for the use, moderation, and management of all public platforms officially used by the project, including but not limited to:

Reviewer: According to this language, it sounds like you're implicitly looking for administrative access to these tools. Is that correct? Would like clarification if not. If so, how do you plan to manage access control to these systems?

Reviewer 2: @robzienert some aspects of this access will be managed through automation that the contributor experience will build. This building phase will create groups to handle things. For example, slack automation. That will not be blanket access but rather controlled by a process that will be based on GitOps, much like we have in the Kubernetes community.

Reviewer 3: Good question @robzienert. Public platform access control is technically managed by the Continuous Delivery Foundation. My observation is that an additional human interface with the CDF is needed in addition to what the CDF provides to facilitate sharing and make sure that access info is communicated to fairly empower contributors while. CDF can help with secure sharing.

> -  In accordance with the code of conduct, establish policies, standards, and procedures for the use, moderation, and management of all public platforms officially used by the project, including but not limited to:

Reviewer 4: I think some examples of the types of policies that might be implemented would be useful.  Depending on the particular policy, I would imagine that we might want to have this SIG work on building consensus and making a proposal, then presenting that to the SC/TOC for comments/final approval.

I don't want to add extra steps for small tactical procedures, but do want to make sure that we are appropriately reviewing changes that have a broader impact (and thus why understanding some of the types of policies would be useful).

Reviewer 3: Nikema is working on adding some more specifics to the doc in response to your suggestion. One thing I'll say generally is, the model we discussed with Rob in his comments is more of an "opt-in" model than an approval gate model. This SIG can establish tools/resources/processes to enable SIGs to manage things or make them more welcoming/equitable. Then next steps would be to pitch or offer those things to the SIGs themselves. If SIGs choose to adopt them, great. Successful SIGs can be a model to non-adopters if the strategies are helpful. That way buy-in happens over time.

I could see an instance in which this SIG might discover something that really should be addressed project-wide and adopted by all SIGs. In those cases, presenting a proposal or RFC to the TOC/SC for approval seems reasonable.

Thoughts @ezimanyi ?

Reviewer 4: This model sounds great to me!  I really like the idea of this SIG helping other SIGs with best practices and tooling and agree (at least initially) that this would be best as opt-in. Just read through the text changes to the RFC which look good.

  - [GitHub Management](https://github.com/spinnaker)
  - [Slack](https://join.spinnaker.io/)
  - [Spinnaker](https://www.youtube.com/channel/UCcxQbw8kT1-FRhFhO2QCetg) YouTube channel
  - Zoom
  - Google Meet
- Set up alerting automation for GitHub events, including new contributors, new issues, and new pull requests.

> - Work with other SIGs and interested parties in the project to execute GitHub tasks where required, see [GitHub Management](https://github.com/spinnaker) for more detail

Reviewer: I'm interested in more detail, but I don't see how the organization homepage would be providing any more clarity / details?

Reviewer 3: This is a good catch. We need to add more clarification and reference prior art from other projects as needed (which is how this oversight came about in the first place, lol).

Reviewer 3: @prophen @ExitoLab this is an easy fix, please check the referenced doc in the K8S project (and consult other porjects if you want), then you can just define whatever you want to say here instead of referencing another link. You can delete this bullet entirely if you want but I think that executing some GitHub mgmt tasks as described might be in scope - your call.

> - Own and execute events that are targeted to the Spinnaker contributor community, including:

Reviewer: Items in this list seem to have direct, conflicting overlaps with the [Summit SIG](https://github.com/spinnaker/governance/tree/master/sig-summit). Is this SIG meant to replace the Summit SIG? Are the leads of the Summit SIG aware of this?

Reviewer 2: To my knowledge this is not a replacement

Reviewer 3: An important distinction here is **contributor** events, which to refers to a contributor day or hackathon event specifically for a project contributor audience. I think that the Spinnaker Summit, which historically has been a conference event, is out of scope. That event is going to be managed by the Linux Foundation events team going forward, in collaboration with the Summit SIG, which will help develop the speaker content program and make platform/program decisions with the LF team in the same way other CDF events are managed.  This SIG can cooperate with the Summit SIG as necessary to add contributor events to the Summit as desired.

IMO: Producing events is a lot of work. Breaking it up this way, by audience, will help us use open processes to to the work collaboratively across stakeholder organizations when planning events. This is LF methodology and prevents one company from dominating the decisions.

@michaelgalloway and @gcomstock, please let us know your thoughts as the co-leads of the SIG : )

Reviewer 5: Hi @dnilasor,

I don't have much to add here. Producing events is a lot of work for sure, and my involvement with this year's summit was thin. I'm all in support. Thank you!

  - Contributor Summit(s) 
  - [Gardening Days and hackathons](https://github.com/spinnaker-hackathon/gardening)
  - Retrospective moderation for other SIGs upon request

>   - Other events, like other SIG face to face events, upon request and consideration

Reviewer: ```suggestion
  - Other events, like other SIG face-to-face events, upon request and consideration
```

> - Strategize, build, and execute on scalable mentoring programs for all contributor levels. These may include:

Reviewer: FYI: https://github.com/spinnaker/governance/pull/51

Reviewer 2: Yes #51 is a good issue. What this meant by mentorship here is 1:1 and group mentoring. The idea is a new user comes into the community and test the waters and when they are ready a mentor will guide them within a SIG. Again, very much like we do in Jenkins and Kubernetes.

  - [Google Summer of Code](https://summerofcode.withgoogle.com/archive/)
  - [Outreachy](https://www.outreachy.org/)

> - New mentoring programs: office hours, meet the contributors, 1:1 pairing, group mentoring, etc.

Reviewer: ```suggestion
- Mentoring programs, such as: office hours, meet the contributors, 1:1 pairing, group mentoring, etc.
```

Reviewer 3: Detail but, it's important to note that this group would be launching a mentorship program (so maybe scope should say that clearly). And then, maintaining it. Perhaps that meaning is the intent of the word "new" and could be clarified further.

Prior art on mentorship program, FWIW: https://github.com/spinnaker/governance/pull/51

- Help onboard new and current contributors into the culture, workflow, and CI of our contributor experience through the [contributor guide](https://spinnaker.io/community/contributing/), other related documentation, contributor summits, and programs tailored to onboarding.

> - Perform issue triage on and maintain the spinnaker/contributor-experience repository.  

Reviewer: 1. What are the planned contents of this repository? 
1. Who will need write access to it?

Reviewer 2: Content of this repository will be automation specific to this SIG (e.g, guides, various automation specific to contribex)
I feel to start the access will just be for leads of the contribEx sig.

- Establish and share metrics to measure project health, community health, and general trends, including:
  - [DevStats](https://spinnaker.devstats.cd.foundation/)
  - the CDF contribution gamification app [[WIP](https://github.com/ExitoLab/spinnaker_gamification_app)]
  - contributor experience survey(s)
  - engagement on project platforms
  - pull request and issue engagement and closure times
- Set community health goals such as improved PR and issue engagement and closure times, and track progress towards their achievement using established metrics.
- Research other OSS projects and consult with their leaders about contributor experience topics to make sure we are always delivering value to our contributors
- Create and maintain subprojects that incentivize community outreach, individual content creation, and welcoming/onboarding of new contributors

#### Cross-project and Externally Facing Processes

We cross-cut all SIGs and working groups to deliver the following processes:

- Deploying Changes:

>  This SIG will create change-deployment policies that balance quick response to the needs of the community and a limit on disruptions for project contributors. We will create policies to socialize and elicit feedback on changes via community channels such as Slack, spinnaker.io, Twitter, mailing lists, SIG meetings, delegating to SIG leads, etc. The amount of notice we provide and the amount of consensus we seek is driven by our estimation of risk. We can estimate risk based on these parameters:

Reviewer 4: Does this mean that this SIG is empowered to establish policies about how all SIGs conduct their business?  (For example, requiring a particular platform for meetings, requiring a particular meeting cadence, etc.).

Similar to the above question, I'm wondering how much oversight/approval we'd expect the TOC/SC to provide in these cases.

Reviewer 4: Updates have provided clarify/resolved this.

- Low-risk changes impact a small number (<3) of SIGs, working groups, or repositories, do not break existing contributor workflows, and are easy to roll back.
- High-risk changes impact a large number (>3) of SIGs, working groups, or repositories, break existing contributor workflows and are not easy to roll back.
- Depending on how wide of an ecosystem change this is, we may also slack, blog, tweet, and use other channels to get the word out.

> - Our standard time box is 72 business hours; however, there may be situations where we need to act quickly but the time period will always be clear and upfront.

Reviewer: 72 business hours is 9 business days. Is this language intentional?

Regardless, I suspect using "business days" is going to lead to problems: Business days are locale-specific, which the Spinnaker organization is not (Canada does not have the same business holidays as the United States, for example).

Reviewer 2: We may need to revisit this time box

Reviewer 3: I think this time box should be revisted after the SIG is founded and determines what the capacity is for response. I recommend deleting this and adding that determination as a goal @ExitoLab @Prophen

- Encourage automation to improve productivity for contributors where it makes sense, and consult with SIG leads on automation workflows.

If we need funding for any reason, we approach [CDF](https://cd.foundation).
CDF may contribute funding to our platforms, processes, and/or programs. They do not play a day-to-day operations role and have bestowed that to our community to run as we see fit. Since they do fund some of our initiatives, this means that they hold owner account privileges on certain platforms like Zoom and Slack. The CX SIG will work to ensure that project maintainers are included as admins.

### Out of scope

> - Code for the testing and CI infrastructure

Reviewer: What about code for the in-scope projects and services? I'm not clear on whether or not this SIG will be writing code in support of its initiatives, or looking for other groups to perform the legwork. For example, there's _some_ code around contributorship today, are you planning to take over these scripts?

In my mind, the SIG should be responsible for the scripts and tooling that comes out of this SIG if it applies to the project as a whole, including scripts that may be a part of the governance repo itself (e.g. https://github.com/spinnaker/governance/tree/master/scripts and more).

If this SIG is _not_ responsible for the code and tooling that will support the processes and policies this SIG sets out to create, I have little optimism in the short- or long-term success of this SIG's overall goals.

Reviewer 2: The contribEx SIG will be writing the code for the initiatives specific to contribex. For example, if there is slack automation, various automation around owners for this SIG/repo, that is where this will live and be handled by the contribex team.

Reviewer 3: @robzienert, can you please link to the existing code around contributorship to for the sake of this conversation? I'd love to request that existing artifacts be referenced specifically in this PR.

Reviewer: I linked to the code in my original comment.

Reviewer 3: Is that the only such existing asset that you would propose this group should maintain, or are there other assets?

Reviewer: I think that's it. All membership management has been a manual affair thus far. Other github actions we are maintaining are covered by other SIGs.

- User community management.
- The contributor experience for repositories not included in the Spinnaker organization

> - We do not create SIGs/working group but can assist in the various community management needs of their micro-communities that would kick off their formation and keep them going.

Reviewer: I'm unclear on how this point co-exists with "user community management" being out of scope.

Reviewer 2: We can help with the formation of SIG's and/or working groups but the idea initiation is up to the users wishing to form so said group. That is different than community management.

Reviewer 3: I see how this wording is unclear and will think about a better suggestion. I think it is trying to say that general user community management is out of scope, as in, this SIG is not wholly or primarily responsible for user community management (which is in scope for the Steering Committee in our docs as written, cc @imosquera). However, this group can assist in ensuring that SIGs to address projects needs are created, and that critical/productive SIGs are maintained as people move in and out of the project. They can do this through enablement and relationship management, which intersects with, and is ideally positioned by, other work described in the charter.

- We are not a code of conduct committee and therefore do not control incident management reporting or decisions; however, our mission includes acting swiftly if there is a clear violation of either our code of conduct or one of the terms of our supported platform of service.

## Roles and Organization Management

> This sig adheres to the Roles and Organization Management outlined in Spinnaker's governance policy and will actively suggest updates to those policies as part of our work.

Reviewer: ```suggestion
This SIG adheres to the Roles and Organization Management outlined in Spinnaker's governance policy and will actively suggest updates to those policies as part of our work.
```

### Additional responsibilities of Chairs

> Chairs SHOULD plan Contributor Experience meetings monthly at the least, and should strive to enable annual face to face interaction as conditions allow. Chairs will also ensure that technical changes from subprojects follow the process change communication guidelines we establish.

Reviewer: ```suggestion
Chairs SHOULD plan Contributor Experience meetings monthly, at the least, and should strive to enable annual face-to-face interaction as conditions allow. Chairs will also ensure that technical changes from subprojects follow the process change communication guidelines we establish.
```
